File: week2/timed_repeat_adder.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int main(int argc, char **argv)
{
    // timing variables
    clock_t start_time, end_time;
    double run_time;

    // start timing
    start_time = clock();

    // creates and initialises the variables
    int i, input, output, multiple;
    i = input = output = multiple = 0;

    // checks if there are the right number of arguments
    if (argc == 3)
    {
        // converts the arguments to integers
        input = atoi(argv[1]);
        multiple = atoi(argv[2]);
    }
    else
    {
        fprintf(stderr, "Incorrect arguments. Usage: repeat_adder [NUM] [NUM]\n");
        fprintf(stderr, "e.g. repeat_adder 2 3\n");
        exit(-1);
    }

    // iterates over all numbers up to input
    for (i = 0; i < input; i++)
    {
        output = output + multiple;
    }

    // end timing
    end_time = clock();

    // calculate runtime
    run_time = (double)(end_time - start_time) / CLOCKS_PER_SEC;

    // prints the output
    printf("Result: %d\n", output);
    printf("Execution time: %f seconds\n", run_time);

    return 0;
}

File: week2/timed_repeat_adder.py
import sys
import time

def main():
    output = 0

    # checks if there are the right number of arguments
    try:
        # converts the first two arguments to integers
        in_arg = int(sys.argv[1])
        multiple = int(sys.argv[2])
    except:
        raise Exception(
            "Incorrect arguments.\n"
            "Usage: python repeat_adder.py [NUM] [NUM]\n"
            "e.g.\npython repeat_adder.py 2 3"
        )

    # start timing
    start_time = time.perf_counter()

    # iterates over all numbers up to the input
    for i in range(in_arg):
        output = output + multiple

    # end timing
    end_time = time.perf_counter()

    # calculate runtime
    run_time = end_time - start_time

    # prints the output
    print(output)
    print(f"Execution time: {run_time:.6f} seconds")


if __name__ == "__main__":
    main()
